E3DC Meter implementation created SIGSEV

### Describe the bug

During normal daily operations, no car attached to any loadpoint, charger in status A i got a SIGSEV in evcc.

```
time="2025-08-12T14:47:07+02:00" level=debug msg="write [{ EMS_REQ_POWER_GRID None <nil> }]"
time="2025-08-12T14:47:07+02:00" level=info msg=disconnected
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1b52fcc]
goroutine 613573 [running]:
github.com/evcc-io/evcc/meter.(*E3dc).CurrentPower(0xc000c0c1b0)
	github.com/evcc-io/evcc/meter/e3dc.go:128 +0x5ac
github.com/cenkalti/backoff/v4.doRetryNotify[...](0xc0013e9ef0?, {0x3b9e8c0, 0xc001417c20}, 0x0, {0x0, 0x0?})
	github.com/cenkalti/backoff/v4@v4.3.0/retry.go:88 +0x12b
github.com/cenkalti/backoff/v4.RetryNotifyWithData[...](...)
	github.com/cenkalti/backoff/v4@v4.3.0/retry.go:54
github.com/cenkalti/backoff/v4.RetryWithData[...](0x0?, {0x3b9e8c0?, 0xc001417c20?})
	github.com/cenkalti/backoff/v4@v4.3.0/retry.go:43 +0x2b
github.com/evcc-io/evcc/core.(*Site).updateGridMeter(0xc0009d0488)
	github.com/evcc-io/evcc/core/site.go:719 +0xad
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.16.0/errgroup/errgroup.go:93 +0x50
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 902
	golang.org/x/sync@v0.16.0/errgroup/errgroup.go:78 +0x93
```



### Steps to reproduce

not sure how to reproduce, happened out of nowhere.

### Configuration details

```yaml
➜  ~ docker exec c96e210c14f8 evcc config --database /data/evcc.db
[main  ] INFO 2025/08/12 16:47:16 config file not found, database-only mode
[db    ] INFO 2025/08/12 16:47:16 using sqlite database: /data/evcc.db
[mqtt  ] INFO 2025/08/12 16:47:16 connecting evcc-124312735 at tcp://192.168.4.5:1883
charger
---
db:3 {Type:template Title: Icon: Product:} map[host:192.168.4.98 id:255 modbus:tcpip template:phoenix-ev-eth]

meter
---
db:5 {Type:template Title: Icon: Product:} map[host:192.168.4.77 key:***** password:***** port:5033 template:e3dc-rscp usage:grid user:*****]
db:6 {Type:template Title: Icon: Product:} map[host:192.168.4.77 key:*****  password:***** port:5033 template:e3dc-rscp usage:pv user:*****]
db:7 {Type:template Title: Icon: Product:} map[host:192.168.4.77 key:*****  password:***** port:5033 template:e3dc-rscp usage:battery user:*****]

vehicle
---
db:1 {Type:template Title: Icon: Product:} map[cache:15m capacity:42 icon:car mode:pv password:***** pin:***** template:fiat title:Fiat 500e user:*****]
db:2 {Type:template Title: Icon: Product:} map[capacity:75 icon:car id:1 mode:pv template:teslamate title:Tesla Model Y]
db:8 {Type:template Title: Icon: Product:Kia Bluelink} map[cache:15m capacity:81.4 icon:car language:en password:***** template:kia title:Kia EV3 user:*****]

loadpoint
---
db:4 {Type: Title: Icon: Product:} map[charger:db:3 circuit:main defaultMode:pv limitSoc:0 maxCurrent:20 meter: mode:pv phasesConfigured:1 planEnergy:0 planPrecondition:0 planTime:0001-01-01T00:00:00Z smartCostLimit:100 soc:map[estimate:false poll:map[interval:3.6e+12 mode:charging]] thresholds:map[disable:map[delay:1.8e+11 threshold:0] enable:map[delay:6e+10 threshold:0]] title:Carport vehicle:]

➜  ~
```

### Log details

```text
time="2025-08-12T14:47:07+02:00" level=debug msg="write [{ EMS_REQ_POWER_GRID None <nil> }]"
time="2025-08-12T14:47:07+02:00" level=info msg=disconnected
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1b52fcc]
goroutine 613573 [running]:
github.com/evcc-io/evcc/meter.(*E3dc).CurrentPower(0xc000c0c1b0)
	github.com/evcc-io/evcc/meter/e3dc.go:128 +0x5ac
github.com/cenkalti/backoff/v4.doRetryNotify[...](0xc0013e9ef0?, {0x3b9e8c0, 0xc001417c20}, 0x0, {0x0, 0x0?})
	github.com/cenkalti/backoff/v4@v4.3.0/retry.go:88 +0x12b
github.com/cenkalti/backoff/v4.RetryNotifyWithData[...](...)
	github.com/cenkalti/backoff/v4@v4.3.0/retry.go:54
github.com/cenkalti/backoff/v4.RetryWithData[...](0x0?, {0x3b9e8c0?, 0xc001417c20?})
	github.com/cenkalti/backoff/v4@v4.3.0/retry.go:43 +0x2b
github.com/evcc-io/evcc/core.(*Site).updateGridMeter(0xc0009d0488)
	github.com/evcc-io/evcc/core/site.go:719 +0xad
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.16.0/errgroup/errgroup.go:93 +0x50
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 902
	golang.org/x/sync@v0.16.0/errgroup/errgroup.go:78 +0x93
```

### What type of operating system or environment does evcc run on?

Linux

### External automation

- [x] I have made sure that no external automation like HomeAssistant or Node-RED is active or accessing any of the mentioned devices when this issue occurs.

### Nightly build

- [x] I have verified that the issue is reproducible with the latest nightly build

### Version

0.207.1

--- a/meter/e3dc.go
+++ b/meter/e3dc.go
@@ -21,7 +21,7 @@ type E3dc struct {
 	dischargeLimit uint32
 	usage          templates.Usage // TODO check if we really want to depend on templates
 	conn           *rscp.Client
-	retry          func(err error) error
+	retry          func() error
 }
 
 func init() {
@@ -90,10 +90,7 @@ func NewE3dc(cfg rscp.ClientConfig, usage templates.Usage, dischargeLimit uint32
 		dischargeLimit: dischargeLimit,
 	}
 
-	m.retry = func(err error) error {
-		if err == nil {
-			return nil
-		}
+	m.retry = func() (err error) {
 		m.conn.Disconnect()
 		m.conn, err = rscp.NewClient(cfg)
 		return err
@@ -118,27 +115,29 @@ func NewE3dc(cfg rscp.ClientConfig, usage templates.Usage, dischargeLimit uint32
 // retryMessage executes a single message request with retry
 func (m *E3dc) retryMessage(msg rscp.Message) (*rscp.Message, error) {
 	result, err := m.conn.Send(msg)
-	if err != nil {
-		if retryErr := m.retry(err); retryErr != nil {
-			return nil, retryErr
-		}
-		// retry the request after successful reconnection
-		return m.conn.Send(msg)
+	if err == nil {
+		return result, nil
 	}
-	return result, nil
+
+	if err := m.retry(); err != nil {
+		return nil, err
+	}
+
+	return m.conn.Send(msg)
 }
 
 // retryMessages executes a multiple message request with retry
 func (m *E3dc) retryMessages(msgs []rscp.Message) ([]rscp.Message, error) {
 	result, err := m.conn.SendMultiple(msgs)
-	if err != nil {
-		if retryErr := m.retry(err); retryErr != nil {
-			return nil, retryErr
-		}
-		// retry the request after successful reconnection
-		return m.conn.SendMultiple(msgs)
+	if err == nil {
+		return result, nil
 	}
-	return result, nil
+
+	if err := m.retry(); err != nil {
+		return nil, err
+	}
+
+	return m.conn.SendMultiple(msgs)
 }
 
 func (m *E3dc) CurrentPower() (float64, error) {

--- a/meter/e3dc.go
+++ b/meter/e3dc.go
@@ -115,25 +115,55 @@ func NewE3dc(cfg rscp.ClientConfig, usage templates.Usage, dischargeLimit uint32
 	return decorateE3dc(m, batterySoc, batteryCapacity, batteryMode, maxacpower), nil
 }
 
+// retryMessage executes a single message request with retry
+func (m *E3dc) retryMessage(requestFunc func() (*rscp.Message, error)) (*rscp.Message, error) {
+	result, err := requestFunc()
+	if err != nil {
+		if retryErr := m.retry(err); retryErr != nil {
+			return nil, retryErr
+		}
+		// retry the request after successful reconnection
+		return requestFunc()
+	}
+	return result, nil
+}
+
+// retryMessages executes a multiple message request with retry
+func (m *E3dc) retryMessages(requestFunc func() ([]rscp.Message, error)) ([]rscp.Message, error) {
+	result, err := requestFunc()
+	if err != nil {
+		if retryErr := m.retry(err); retryErr != nil {
+			return nil, retryErr
+		}
+		// retry the request after successful reconnection
+		return requestFunc()
+	}
+	return result, nil
+}
+
 func (m *E3dc) CurrentPower() (float64, error) {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 
 	switch m.usage {
 	case templates.UsageGrid:
-		res, err := m.conn.Send(*rscp.NewMessage(rscp.EMS_REQ_POWER_GRID, nil))
-		if e := m.retry(err); e != nil {
-			return 0, e
+		res, err := m.retryMessage(func() (*rscp.Message, error) {
+			return m.conn.Send(*rscp.NewMessage(rscp.EMS_REQ_POWER_GRID, nil))
+		})
+		if err != nil {
+			return 0, err
 		}
 		return rscpValue(*res, cast.ToFloat64E)
 
 	case templates.UsagePV:
-		res, err := m.conn.SendMultiple([]rscp.Message{
-			*rscp.NewMessage(rscp.EMS_REQ_POWER_PV, nil),
-			*rscp.NewMessage(rscp.EMS_REQ_POWER_ADD, nil),
+		res, err := m.retryMessages(func() ([]rscp.Message, error) {
+			return m.conn.SendMultiple([]rscp.Message{
+				*rscp.NewMessage(rscp.EMS_REQ_POWER_PV, nil),
+				*rscp.NewMessage(rscp.EMS_REQ_POWER_ADD, nil),
+			})
 		})
-		if e := m.retry(err); e != nil {
-			return 0, e
+		if err != nil {
+			return 0, err
 		}
 
 		values, err := rscpValues(res, cast.ToFloat64E)
@@ -144,9 +174,11 @@ func (m *E3dc) CurrentPower() (float64, error) {
 		return values[0] - values[1], nil
 
 	case templates.UsageBattery:
-		res, err := m.conn.Send(*rscp.NewMessage(rscp.EMS_REQ_POWER_BAT, nil))
-		if e := m.retry(err); e != nil {
-			return 0, e
+		res, err := m.retryMessage(func() (*rscp.Message, error) {
+			return m.conn.Send(*rscp.NewMessage(rscp.EMS_REQ_POWER_BAT, nil))
+		})
+		if err != nil {
+			return 0, err
 		}
 		pwr, err := rscpValue(*res, cast.ToFloat64E)
 		if err != nil {
@@ -164,9 +196,11 @@ func (m *E3dc) batterySoc() (float64, error) {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 
-	res, err := m.conn.Send(*rscp.NewMessage(rscp.EMS_REQ_BAT_SOC, nil))
-	if e := m.retry(err); e != nil {
-		return 0, e
+	res, err := m.retryMessage(func() (*rscp.Message, error) {
+		return m.conn.Send(*rscp.NewMessage(rscp.EMS_REQ_BAT_SOC, nil))
+	})
+	if err != nil {
+		return 0, err
 	}
 
 	return rscpValue(*res, cast.ToFloat64E)
@@ -176,36 +210,32 @@ func (m *E3dc) setBatteryMode(mode api.BatteryMode) error {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 
-	var (
-		res []rscp.Message
-		err error
-	)
-
+	var messages []rscp.Message
 	switch mode {
 	case api.BatteryNormal:
-		res, err = m.conn.SendMultiple([]rscp.Message{
+		messages = []rscp.Message{
 			e3dcDischargeBatteryLimit(false, 0),
 			e3dcBatteryCharge(0),
-		})
-
+		}
 	case api.BatteryHold:
-		res, err = m.conn.SendMultiple([]rscp.Message{
+		messages = []rscp.Message{
 			e3dcDischargeBatteryLimit(true, m.dischargeLimit),
 			e3dcBatteryCharge(0),
-		})
-
+		}
 	case api.BatteryCharge:
-		res, err = m.conn.SendMultiple([]rscp.Message{
+		messages = []rscp.Message{
 			e3dcDischargeBatteryLimit(false, 0),
 			e3dcBatteryCharge(50000), // max. 50kWh
-		})
-
+		}
 	default:
 		return api.ErrNotAvailable
 	}
 
-	if e := m.retry(err); e != nil {
-		return e
+	res, err := m.retryMessages(func() ([]rscp.Message, error) {
+		return m.conn.SendMultiple(messages)
+	})
+	if err != nil {
+		return err
 	}
 
 	return rscpError(res...)

--- a/meter/e3dc.go
+++ b/meter/e3dc.go
@@ -116,27 +116,27 @@ func NewE3dc(cfg rscp.ClientConfig, usage templates.Usage, dischargeLimit uint32
 }
 
 // retryMessage executes a single message request with retry
-func (m *E3dc) retryMessage(requestFunc func() (*rscp.Message, error)) (*rscp.Message, error) {
-	result, err := requestFunc()
+func (m *E3dc) retryMessage(msg rscp.Message) (*rscp.Message, error) {
+	result, err := m.conn.Send(msg)
 	if err != nil {
 		if retryErr := m.retry(err); retryErr != nil {
 			return nil, retryErr
 		}
 		// retry the request after successful reconnection
-		return requestFunc()
+		return m.conn.Send(msg)
 	}
 	return result, nil
 }
 
 // retryMessages executes a multiple message request with retry
-func (m *E3dc) retryMessages(requestFunc func() ([]rscp.Message, error)) ([]rscp.Message, error) {
-	result, err := requestFunc()
+func (m *E3dc) retryMessages(msgs []rscp.Message) ([]rscp.Message, error) {
+	result, err := m.conn.SendMultiple(msgs)
 	if err != nil {
 		if retryErr := m.retry(err); retryErr != nil {
 			return nil, retryErr
 		}
 		// retry the request after successful reconnection
-		return requestFunc()
+		return m.conn.SendMultiple(msgs)
 	}
 	return result, nil
 }
@@ -147,20 +147,16 @@ func (m *E3dc) CurrentPower() (float64, error) {
 
 	switch m.usage {
 	case templates.UsageGrid:
-		res, err := m.retryMessage(func() (*rscp.Message, error) {
-			return m.conn.Send(*rscp.NewMessage(rscp.EMS_REQ_POWER_GRID, nil))
-		})
+		res, err := m.retryMessage(*rscp.NewMessage(rscp.EMS_REQ_POWER_GRID, nil))
 		if err != nil {
 			return 0, err
 		}
 		return rscpValue(*res, cast.ToFloat64E)
 
 	case templates.UsagePV:
-		res, err := m.retryMessages(func() ([]rscp.Message, error) {
-			return m.conn.SendMultiple([]rscp.Message{
-				*rscp.NewMessage(rscp.EMS_REQ_POWER_PV, nil),
-				*rscp.NewMessage(rscp.EMS_REQ_POWER_ADD, nil),
-			})
+		res, err := m.retryMessages([]rscp.Message{
+			*rscp.NewMessage(rscp.EMS_REQ_POWER_PV, nil),
+			*rscp.NewMessage(rscp.EMS_REQ_POWER_ADD, nil),
 		})
 		if err != nil {
 			return 0, err
@@ -174,9 +170,7 @@ func (m *E3dc) CurrentPower() (float64, error) {
 		return values[0] - values[1], nil
 
 	case templates.UsageBattery:
-		res, err := m.retryMessage(func() (*rscp.Message, error) {
-			return m.conn.Send(*rscp.NewMessage(rscp.EMS_REQ_POWER_BAT, nil))
-		})
+		res, err := m.retryMessage(*rscp.NewMessage(rscp.EMS_REQ_POWER_BAT, nil))
 		if err != nil {
 			return 0, err
 		}
@@ -196,9 +190,7 @@ func (m *E3dc) batterySoc() (float64, error) {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 
-	res, err := m.retryMessage(func() (*rscp.Message, error) {
-		return m.conn.Send(*rscp.NewMessage(rscp.EMS_REQ_BAT_SOC, nil))
-	})
+	res, err := m.retryMessage(*rscp.NewMessage(rscp.EMS_REQ_BAT_SOC, nil))
 	if err != nil {
 		return 0, err
 	}
@@ -231,9 +223,7 @@ func (m *E3dc) setBatteryMode(mode api.BatteryMode) error {
 		return api.ErrNotAvailable
 	}
 
-	res, err := m.retryMessages(func() ([]rscp.Message, error) {
-		return m.conn.SendMultiple(messages)
-	})
+	res, err := m.retryMessages(messages)
 	if err != nil {
 		return err
 	}

